engine/minskyTensorOps.cc

```diff
@@ -315,26 +320,23 @@ namespace minsky
         if (arg1->hypercube().dims()[arg1->rank()-1]!=arg2->hypercube().dims()[0] && (arg1->rank()<arg2->rank()))  // do not allow products of the type nx1 dot nxm
           throw std::runtime_error("inner dimensions of tensors do not match");
         
-        // shape of output tensor should match outer dimensions of the two rank<2 tensors originally multiplied by one another     
-        if (arg1->rank()>1 && arg2->rank()>1) {
-          vector<XVector>&& outerXVs{arg1->hypercube().xvectors[0],arg2->hypercube().xvectors[arg2->rank()-1]};
-          cachedResult.hypercube(outerXVs);
-        }
-        else if (arg1->rank()>1 && arg2->rank()<2) { //case mxn dot nx1
-          vector<XVector>&& outerXVs{arg2->hypercube().xvectors[0]};
-          cachedResult.hypercube(outerXVs);
-        }
-        else {
-			//cachedResult.hypercube({});  // causes segfault: 0x00005555556d3926 in boost::any::~any (this=0x55555723f850, __in_chrg=<optimized out>) at /usr/local/include/boost/any.hpp:79 delete content; 
-			
-			// also causes the above segfault
-            auto hc=arg2->hypercube();
-            if (rank()==0) return;
-            
-            auto& xv=hc.xvectors[0];
-            xv.erase(xv.end()-xv.size()-2, xv.end());
-            cachedResult.hypercube(move(hc));
-		}
+        auto xv1=arg1->hypercube().xvectors, xv2=arg2->hypercube().xvectors;
+        Hypercube hc;
+        hc.xvectors.insert(hc.xvectors.begin(), xv1.begin(), xv1.end()-1);
+        hc.xvectors.insert(hc.xvectors.begin(), xv2.begin()+1, xv2.end());
+        cachedResult.hypercube(move(hc));
+        
+        auto idx1=arg1->index(), idx2=arg2->index();
+        set<size_t> newIdx;        
+        for (auto& i: idx1)
+          for (auto& j: idx2)
+          {
+            // strip of any indices outside the output range
+            auto t=ssize_t(j)-ssize_t(i);
+            if (t>=0 && t<ssize_t(size()))  newIdx.insert(t);
+          }
+          
+        cachedResult.index(Index(newIdx));           
```

Comment:
For now, computeTensor is not sparse aware, so the only way this can be consistent is to not set the index of cachedResult. I suggest you commit this "densifying" inner product for now (once you fix the above algorithm for n>1), and leave correct sparse handling for a later PR.

Reply:
> For now, computeTensor is not sparse aware, so the only way this can be consistent is to not set the index of cachedResult

Okay, this answers my question in the previous reply. Thanks.